packages/agreement-process/src/services/readModelService.ts

```diff
@@ -289,106 +295,135 @@ async function searchTenantsByName(
+  const delegations = readModelRepository.delegations;
+  return {
+    async getAgreements(
+      filters: AgreementQueryFilters,
+      limit: number,
+      offset: number
+    ): Promise<ListResult<Agreement>> {
+      const { producerId, ...filtersWithoutProducerId } = filters;
+      const producerIds = producerId
+        ? Array.isArray(producerId)
+          ? producerId
+          : [producerId]
+        : [];
+
+      const agreementsData = await agreements
         .aggregate(
-          [...aggregationPipeline, { $skip: offset }, { $limit: limit }],
-          { allowDiskUse: true }
+          [
+            getAgreementsFilters(filtersWithoutProducerId),
+            ...getDelegateAgreementsFilters(producerIds),
+          ],
+          {
+            allowDiskUse: true,
+          }
         )
         .toArray();
 
-      const result = z.array(Agreement).safeParse(data.map((d) => d.data));
+      const eserviceIds = agreementsData.map(
+        (agreement) => agreement.data.eserviceId
+      );
+      const eservicesData = await eservices
+        .find({ "data.id": { $in: eserviceIds } })
+        .toArray();
+
+      const eservicesMap = new Map(
+        eservicesData.map((eservice) => [eservice.data.id, eservice.data])
+      );
+
+      const combinedData: Array<{
+        agreement: AgreementReadModel;
+        eservice: EServiceReadModel;
+      }> = agreementsData.flatMap((agreement) => {
+        const eservice = eservicesMap.get(agreement.data.eserviceId);
+        return eservice ? [{ agreement: agreement.data, eservice }] : [];
+      });
+
+      const filteredData = filters.showOnlyUpgradeable
+        ? combinedData.filter((cb) => {
+            const currentDescriptor = cb.eservice.descriptors.find(
+              (descr) => descr.id === cb.agreement.descriptorId
+            );
+            const upgradableDescriptor = cb.eservice.descriptors.filter(
+              (upgradable: DescriptorReadModel) => {
+                // Since the dates are optional, if they are undefined they are set to a very old date
+                const currentPublishedAt =
+                  currentDescriptor?.publishedAt ?? new Date(0);
+                const upgradablePublishedAt =
+                  upgradable.publishedAt ?? new Date(0);
+                return (
+                  upgradablePublishedAt > currentPublishedAt &&
+                  (upgradable.state === descriptorState.published ||
+                    upgradable.state === descriptorState.suspended)
+                );
+              }
+            );
+            return upgradableDescriptor.length > 0;
+          })
+        : combinedData;
+
+      const data = filteredData
+        .slice(offset, offset + limit)
+        .sort((a, b) =>
+          a.eservice.name
+            .toLowerCase()
+            .localeCompare(b.eservice.name.toLowerCase())
+        );
+
+      const result = z.array(Agreement).safeParse(data.map((d) => d.agreement));
```

Comment:
This change is a rewrite of this PR https://github.com/pagopa/interop-be-monorepo/pull/813 to include delegations